[Coro] Amortize debug info processing cost in CoroSplit

artempyanykh opened the pull request:
More details about this stack are [in this topic on discourse](https://discourse.llvm.org/t/rfc-amortizing-debug-info-processing-cost-in-corosplit/81275). 

TL;DR: In large modules with a bunch of coroutines the cost of debug info metadata processing can get very high (e.g. adding over a minute to the compile time). This stack does some refactoring around function cloning to significantly speed up the pass.

This makes CoroSplit pass almost as fast with -g2 as it is with -g1 on
the sample cpp file used with other parts of this stack:

|                 | Baseline | IdentityMD set | Prebuilt CommonDI | MetadataPred (fin) |
|-----------------|----------|----------------|-------------------|--------------------|
| CoroSplitPass   | 306ms    | 221ms          | 68ms              | 3.8ms              |
| CoroCloner      | 101ms    | 72ms           | 0.5ms             | 0.5ms              |
| CollectCommonDI | -        | -              | 63ms              | -                  |
| Speed up        | 1x       | 1.4x           | 4.5x              | 80x                |

Each commit is individually buildable and reviewable. I can extract them into separate PRs if the overall direction makes sense. 

-----

The final version has been merged in a series of PRs (❗️marks the salient ones):
1. #112948
2. #112976
3. #114537
4. #118620
5. #118621
6. #118622
7. #118623 
8. #118624
9. #118625
10. ❗️ #118627
11. #118628
12. #129143
13. #129144
14. #129145
15. #129146
16. ❗️ #129147
17. ❗️ #129148
18. #129149
19. #129150
20. #129151
21. #129152
22. #129153
23. #129154

artempyanykh committed:
[NFC][Utils] Extract BuildDebugInfoMDMap from CloneFunctionInto

Summary:
Extract the logic to build up a metadata map to use in metadata cloning
into a separate function.

Test Plan:
ninja check-llvm-unit check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118622, branch: users/artempyanykh/fast-coro-upstream/3

artempyanykh committed:
[NFC][Utils] Extract CloneFunctionMetadataInto from CloneFunctionInto

Summary:
The new API expects the caller to populate the VMap. We need it this way
for a subsequent change around coroutine cloning.

Test Plan:
ninja check-llvm-unit check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118623, branch: users/artempyanykh/fast-coro-upstream/4

artempyanykh committed:
[NFC][Utils] Extract CloneFunctionBodyInto from CloneFunctionInto

Summary:
This and previously extracted `CloneFunction*Into` functions will be used in later diffs.

Test Plan:
ninja check-llvm-unit check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118624, branch: users/artempyanykh/fast-coro-upstream/5

artempyanykh committed:
[NFC][Utils] Eliminate DISubprogram set from BuildDebugInfoMDMap

Summary:
Previously, we'd add all SPs distinct from the cloned one into a set.
Then when cloning a local scope we'd check if it's from one of those
'distinct' SPs by checking if it's in the set. We don't need to do that.
We can just check against the cloned SP directly and drop the set.

Test Plan:
ninja check-llvm-unit check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118625, branch: users/artempyanykh/fast-coro-upstream/6

artempyanykh committed:
[Utils] Identity map module-level debug info on first use in CloneFunction*

Summary:
To avoid cloning module-level debug info (owned by the module rather
than the function), CloneFunction implementation used to eagerly
identity map such debug info into ValueMap's MD map. In larger modules
with meaningful volume of debug info this gets very expensive.

By passing such debug info metadata via an IdentityMD set for the
ValueMapper to map on first use, we get several benefits:

1. Mapping metadata is not cheap, particularly because of tracking. When
   cloning a Function we identity map lots of global module-level
   metadata to avoid cloning it, while only a fraction of it is actually
   used by the function. Mapping on first use is a lot faster for
   modules with meaningful amount of debug info.

2. Eagerly identity mapping metadata makes it harder to cache
   module-level data (e.g. a set of metadata nodes in a \a DICompileUnit).
   With this patch we can cache certain module-level metadata
   calculations to speed things up further.

Anecdata from compiling a sample cpp file with full debug info shows that this moderately speeds up
CoroSplitPass which is one of the heavier users of cloning:

|                 | Baseline | IdentityMD set |
|-----------------|----------|----------------|
| CoroSplitPass   | 306ms    | 221ms          |
| CoroCloner      | 101ms    | 72ms           |
| Speed up        | 1x       | 1.4x           |

Test Plan:
ninja check-llvm-unit
ninja check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118627, branch: users/artempyanykh/fast-coro-upstream/8

artempyanykh committed:
[Analysis] Add DebugInfoCache analysis

Summary:
The analysis simply primes and caches DebugInfoFinders for each DICompileUnit in a module. This
allows (future) callers like CoroSplitPass to compute global debug info metadata (required for
coroutine function cloning) much faster. Specifically, pay the price of DICompileUnit processing
only once per compile unit, rather than once per coroutine.

Test Plan:
Added a smoke test for the new analysis
ninja check-llvm-unit check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118629, branch: users/artempyanykh/fast-coro-upstream/10

artempyanykh committed:
[Coro] Use DebugInfoCache to speed up cloning in CoroSplitPass

Summary:
We can use a DebugInfoFinder from DebugInfoCache which is already primed on a compile unit to speed
up collection of module-level debug info.

The pass could likely be another 2x+ faster if we avoid rebuilding the set of global debug
info. This needs further massaging of CloneFunction and ValueMapper, though, and can be done
incrementally on top of this.

Comparing performance of CoroSplitPass at various points in this stack, this is anecdata from a sample
cpp file compiled with full debug info:
|                 | Baseline | IdentityMD set | Prebuilt CommonDI | Cached CU DIFinder (cur.) |
|-----------------|----------|----------------|-------------------|---------------------------|
| CoroSplitPass   | 306ms    | 221ms          | 68ms              | 17ms                      |
| CoroCloner      | 101ms    | 72ms           | 0.5ms             | 0.5ms                     |
| CollectGlobalDI | -        | -              | 63ms              | 13ms                      |
| Speed up        | 1x       | 1.4x           | 4.5x              | 18x                       |

Test Plan:
ninja check-llvm-unit
ninja check-llvm

Compiled a sample cpp file with time trace to get the avg. duration of the pass and inner scopes.

stack-info: PR: https://github.com/llvm/llvm-project/pull/118630, branch: users/artempyanykh/fast-coro-upstream/11

adrian-prantl commented on llvm/lib/Analysis/DebugInfoCache.cpp:
```diff
@@ -0,0 +1,47 @@
+//===- llvm/Analysis/DebugInfoCache.cpp - debug info cache ---*- C++ -*-===//
```
nit: the -*- C++ -*- only makes sense for .h files.

adrian-prantl commented on llvm/include/llvm/Analysis/DebugInfoCache.h:
```diff
@@ -0,0 +1,43 @@
+#include "llvm/IR/PassManager.h"
+
+namespace llvm {
+class DebugInfoCache {
```
Can you add doxygen comments explaining what exactly is being cached and what the use-case for this is?

artempyanykh replied:
Thanks @adrian-prantl, will do!

adrian-prantl reviewed (commented):
This sounds like generally a good idea and the implementation looks clean as well.
What are the behavior changes we expect from this patch series?

artempyanykh commented:
> This sounds like generally a good idea and the implementation looks clean as well.

Thanks for reviewing @adrian-prantl!

> What are the behavior changes we expect from this patch series?

* I expect no functional changes, only faster compile times for C++ code with coroutines. 
* The new DebugInfoCache analysis is not conditioned on the presence of coroutines in a module but it's fast and the analysis results can be used to speed up function cloning in other places than just coroutines.
* I ran a compile time benchmark on a few thousand internal source files:
    * There were no perf regressions, code without coroutines had neutral perf impact.
    * Source files that use coroutines compiled 7% faster on avg, but there were many cases where it was 25%, 50% and even 65% faster. The larger the module is and the more coroutine it has, the greater the speed up.

felipepiovezan commented:
Thanks for working on this and for making each commit reviewable / commitable on its own, that's really great work!
I will start looking at them asap, wish I had seen this sooner! :)

artempyanykh commented:
Thanks @felipepiovezan! There's some commentary around the commits in [this discourse topic](https://discourse.llvm.org/t/rfc-amortizing-debug-info-processing-cost-in-corosplit/81275), in case it's useful.

artempyanykh commented:
@adrian-prantl, @felipepiovezan have you had a chance to take a closer look? Could you advise on what would be the best way to proceed with reviewing this PR?

adrian-prantl reviewed (commented):
@felipepiovezan Can you apply this patch to the LLVM in swiftlang and run the tests with it as a sanity check?

felipepiovezan commented:
Hi @artempyanykh, while I follow Adrian's suggestions, you could extract the first commit into a separate PR and we'll start reviwing/merging the NFC stuff one by one

felipepiovezan commented on llvm/lib/Transforms/Utils/CloneFunction.cpp:
```diff
@@ -86,28 +86,14 @@ BasicBlock *llvm::CloneBasicBlock(const BasicBlock *BB, ValueToValueMapTy &VMap,
-
-  // Copy all attributes other than those stored in the AttributeList.  We need
-  // to remap the parameter indices of the AttributeList.
+// Copy all attributes other than those stored in the AttributeList.  We need
```
I'm not sure this comment is making a lot of sense as a function-level comment. To be honest even the original comment was confusing: which AttributeList?  If you know the answer to this, could you take the chance provided by this NFC commit to improve the comment?

felipepiovezan replied:
This second commit could also use a brief commit message describing why this change is useful

artempyanykh replied:
I think the comment refers to the `AttributeList` field (the only such) of a Function which has e.g. parameters and return value attributes. I updated the comment in #112976 to (hopefully) make it a bit clearer.

felipepiovezan commented on llvm/lib/Transforms/Utils/CloneFunction.cpp:
```diff
@@ -135,6 +135,29 @@ void llvm::CloneFunctionAttributesInto(Function *NewFunc,
+  DISubprogram *SPClonedWithinModule = nullptr;
+  if (Changes < CloneFunctionChangeType::DifferentModule) {
+    SPClonedWithinModule = F.getSubprogram();
+  }
```
LLVM's coding guidelines prescribe not using `{}` around single line blocks

felipepiovezan commented on llvm/lib/Transforms/Utils/CloneFunction.cpp:
```diff
@@ -135,6 +135,29 @@ void llvm::CloneFunctionAttributesInto(Function *NewFunc,
                          OldAttrs.getRetAttrs(), NewArgAttrs));
 }
 
+DISubprogram *llvm::ProcessSubprogramAttachment(const Function &F,
```
Could you add documentation for this function? The term "Process" is not very descriptive. IIUC this function is visiting instructions / SPs  and collecting debug information?

felipepiovezan commented on llvm/include/llvm/Transforms/Utils/Cloning.h:
```diff
@@ -205,6 +205,11 @@ void CloneAndPruneFunctionInto(Function *NewFunc, const Function *OldFunc,
                                ClonedCodeInfo *CodeInfo = nullptr);
 
+/// Process debug information from function's subprogram attachment.
+DISubprogram *ProcessSubprogramAttachment(const Function &F,
```
Could you elaborate on "Process"? This is more about collecting debug info, right?

artempyanykh replied:
That's right. This function hydrates the passed DebugInfoFinder by visiting relevant components of the function's subprogram attachment. I'll update the name/comment when we get to extracting that commit in a separate PR.

felipepiovezan replied:
Sounds good, now that the other two PRs are open, let's proceed with this commit!

artempyanykh replied:
Great, yes! @felipepiovezan a silly question -- now that those PRs are approved, how/when do they get merged? Stacking unmerged PRs on top of each other doesn't have the best UX in Github.

felipepiovezan replied:
So I think the best option here is to merge the two PRs that are approved (https://github.com/llvm/llvm-project/pull/112976 and https://github.com/llvm/llvm-project/pull/112948) . Both of those stand on their own and don't depend on each other, so it's fine to merge in w/e order, right?

And then you can grab the third commit (or more, if the subsequent commits are also independent of each other) of _this_ PR and open a new PR (or multiple PRs, if you did the () before), so it will look fine on top of `main`. My understanding is that the fourth commit depends on the third though.

Regarding what to do about this umbrella PR in the mean time, you have two options: leave this as is (with the understanding that it is not meant to be merged, merely serve as a reference for all the work left to do), or rebase on top of main, which will get rid of the two-already merged commits and force push to your branch `artempyanykh:fast-coro-upstream`, which will mean this PR serves a reference of the work that's left to be merged, also with the understand that we're not merging it.

With all that stack, we'd never have to stack unmerged PRs on top of each other, does that make sense?

felipepiovezan replied:
By the way, do you have commit access?

artempyanykh replied:
> So I think the best option here is to merge the two PRs that are approved (https://github.com/llvm/llvm-project/pull/112976 and https://github.com/llvm/llvm-project/pull/112948) . Both of those stand on their own and don't depend on each other, so it's fine to merge in w/e order, right?

That's right, yes.

> And then you can grab the third commit (or more, if the subsequent commits are also independent of each other) of this PR and open a new PR (or multiple PRs, if you did the () before), so it will look fine on top of main. My understanding is that the fourth commit depends on the third though.

After the first 2 are merged, extracting the third one in a separate PR is straightforward. With the rest of the stack there's quite a bit of sequencing, so we may need to rinse and repeat.

> rebase on top of main, which will get rid of the two-already merged commits and force push to your branch artempyanykh:fast-coro-upstream, which will mean this PR serves a reference of the work that's left to be merged, also with the understand that we're not merging it

+1 to this. I was planning to keep rebasing this stack anyway to make sure things keep working end-to-end while the chunks are being extracted and merged.

> By the way, do you have commit access?

No, not sure what the process for this is. Why?

felipepiovezan replied:
Sorry for the delay, the LLVM dev conference happened last week and we were mostly out.

> No, not sure what the process for this is. Why?

Just wondering if you needed help merging the open PRs. I've merged the first two for you.
The process for getting commit access is described here: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access
If you don't want to do that, you can always let the reviewers know you need someone to press the merge button for you

artempyanykh replied:
> Just wondering if you needed help merging the open PRs. I've merged the first two for you.

Thank you @felipepiovezan, appreciate the help! I also emailed Chris and got the commit rights.

I'll go ahead and extract the next PR.

artempyanykh commented:
@felipepiovezan I extracted the first 2 commits into #112948 and #112976.

artempyanykh commented:
I extracted the next commit into #114537 
Also, rebased this PR

artempyanykh commented:
With #114537 out, there’s quite a bit of sequencing in the remaining 12 commits. I can combine some of the remaining commits into larger PRs if it makes the review logistics a bit easier. @felipepiovezan @adrian-prantl please let me know what you think.

adrian-prantl commented:
> With https://github.com/llvm/llvm-project/pull/114537 out, there’s quite a bit of sequencing in the remaining 12 commits. I can combine some of the remaining commits into larger PRs if it makes the review logistics a bit easier.

On the contrary, 12 PRs with a trivial NFC refactor will be easier to review than 1 large one.

artempyanykh committed:
[Coro] Prebuild a module-level debug info set and share it between all coroutine clones

Summary:
CoroCloner, by calling into CloneFunctionInto, does a lot of repeated
work priming DIFinder and building a list of common module-level debug
info metadata. For programs compiled with full debug info this can get
very expensive.

This diff builds the data once and shares it between all clones.

Anecdata for a sample cpp source file compiled with full debug info:

|                 | Baseline | IdentityMD set | Prebuilt CommonDI (cur.) |
|-----------------|----------|----------------|--------------------------|
| CoroSplitPass   | 306ms    | 221ms          | 68ms                     |
| CoroCloner      | 101ms    | 72ms           | 0.5ms                    |
| CollectGlobalDI | -        | -              | 63ms                     |
| Speed up        | 1x       | 1.4x           | 4.5x                     |

Note that CollectCommonDebugInfo happens once *per coroutine* rather than per clone.

Test Plan:
ninja check-llvm-unit
ninja check-llvm

Compiled a sample internal source file, checked time trace output for scope timings.

stack-info: PR: https://github.com/llvm/llvm-project/pull/118628, branch: users/artempyanykh/fast-coro-upstream/9

artempyanykh commented:
Rebased on trunk

artempyanykh commented:
> On the contrary, 12 PRs with a trivial NFC refactor will be easier to review than 1 large one.

I agree. Just being mindful of sequencing and iteration time. After the review of #114537 is complete, I'll try using [spr](https://github.com/spacedentist/spr) or [stack-pr](https://github.com/modularml/stack-pr) to parallelize the rest of the stack a bit.

felipepiovezan commented:
Hi @artempyanykh, just checking  the status of my reviews after the Thanksgiving break; AFAICT there are no new patches in this series, right?

artempyanykh commented:
Thanks for checking @felipepiovezan! I sent the rest of the stack as PRs via [stack-pr](https://github.com/modularml/stack-pr): the root is at #118620. They still need to land in order, but could be reviewed separately/concurrently. Bottom half of the stack are refactorings. More interesting changes start at #118627.

artempyanykh commented:
Rebase with latest changes.

artempyanykh commented:
This has now been fully merged. I updated the description with references to relevant PRs. The final implementation turned out faster than the initial one: in terms of the CoroSplit pass runtime on a sample cpp file with `-g2`: 306ms -> 3.8ms @ 80x now vs 306ms -> 17ms @ 18x which is almost as fast as `-g1`.

Thanks @felipepiovezan, @TylerNowicki, and @adrian-prantl for reviews!
